skorch/tests/test_doctor.py

```diff
@@ -0,0 +1,742 @@
+        logs = doctor.param_update_logs_
+        assert set(logs.keys()) == {'module', 'criterion'}
+
+        # nothing logged for criterion
+        assert logs['criterion'] == []
+
+        logs_module = logs['module']
+        # 3 epochs, 2 batches per epoch
+        assert len(logs_module) == 6
+
+        # each batch has weights and biases for lin0 & lin1
+        expected = {'lin0.weight', 'lin0.bias', 'lin1.weight', 'lin1.bias'}
+        for batch in logs_module:
+            assert set(batch.keys()) == expected
+
+    def test_param_update_logs_values(self, doctor):
+        logs= doctor.param_update_logs_['module']
+        assert all(np.isscalar(val) for d in logs for val in d.values())
+
+        # for the very first batch, before any update, we actually know that the
+        # updates must be 0 because the gradients are 0.
+        batch0 = logs[0]
+        assert np.isclose(batch0['lin0.weight'], 0)
+        assert np.isclose(batch0['lin0.bias'], 0)
+
+    def test_param_update_logs_not_all_identical(self, doctor):
+        # make sure that values are not just all identical, using a large
+        # tolerance to exclude small deviations
+        logs = doctor.param_update_logs_['module']
+
+        logs_lin0_weight = [log['lin0.weight'] for log in logs]
+        for upd0, upd1 in itertools.combinations(logs_lin0_weight, r=2):
+            assert not np.isclose(upd0, upd1, rtol=1e-3)
+
+        logs_lin0_bias = [log['lin0.bias'] for log in logs]
+        for upd0, upd1 in itertools.combinations(logs_lin0_bias, r=2):
+            assert not np.isclose(upd0, upd1, rtol=1e-3)
+
+        logs_lin1_weight = [log['lin1.weight'] for log in logs]
+        for upd0, upd1 in itertools.combinations(logs_lin1_weight, r=2):
+            assert not np.isclose(upd0, upd1, rtol=1e-3)
+
+        logs_lin1_bias = [log['lin1.bias'] for log in logs]
+        for upd0, upd1 in itertools.combinations(logs_lin1_bias, r=2):
+            assert not np.isclose(upd0, upd1, rtol=1e-3)
+
+    def test_hooks_cleaned_up_after_fit(self, doctor, data):
+        # make sure that the hooks are cleaned up by checking that no more logs
+        # are written when continuing to fit the net
+        num_activation_logs_before = len(doctor.activation_logs_['module'])
+        num_gradient_logs_before = len(doctor.gradient_logs_['module'])
+
+        net = doctor.net
+        net.partial_fit(*data)
+
+        num_activation_logs_after = len(doctor.activation_logs_['module'])
+        num_gradient_logs_after = len(doctor.gradient_logs_['module'])
+
+        assert num_activation_logs_before == num_activation_logs_after
+        assert num_gradient_logs_before == num_gradient_logs_after
+
+    def test_callbacks_cleaned_up_after_fit(self, doctor, net_cls, module_cls):
+        # make sure that the callbacks are the same before and after, this is
+        # important because SkorchDoctor will temporarily add a callback
```

Comment:
I'm trying to find where `SkorchDoctor` adds a callback. May you point to the code where this happens?